### PR TITLE
Feat 3.1 design improvement

### PR DESCRIPTION
**Pull request by JaydenLiang**

### CODE REVIEW

**Review by klc178 — CHANGES_REQUESTED** [review]

File: autoscale-core.ts [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+            // egress traffic route table is set in in EgressTrafficRouteTableList
+            await this.handleEgressTrafficRoute();
+        }
+
+        // need to update the health check record again due to primary ip changes.
+        if (needToUpdateHealthCheckRecord) {
+            this.env.targetHealthCheckRecord.primaryIp = updatedPrimaryIp;
+            await this.platform.updateHealthCheckRecord(this.env.targetHealthCheckRecord);
+            response = JSON.stringify({
+                'master-ip': updatedPrimaryIp
```
[the remark this answers is not in the document]
Comment (JaydenLiang) [reply autoscale-core.ts:364]: this change needs to be supported on FOS side first. skip it for now.

File: fortigate-autoscale/assets/configset/baseconfig [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    unset secondary
+end
+config system global
+    set admin-sport {ADMIN_PORT}
+end
+config system auto-scale
+    set status enable
+    set sync-interface "{SYNC_INTERFACE}"
+    set hb-interval {HEART_BEAT_INTERVAL}
+    set role master
```
[the remark this answers is not in the document]
Comment (JaydenLiang) [reply fortigate-autoscale/assets/configset/baseconfig:12]: This change on FOS CLI config needs to be supported on FOS side first. skip it for now.

File: autoscale-setting.ts [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+    EnableInternalElb = 'enable-internal-elb',
+    EnableNic2 = 'enable-second-nic',
+    EnableVmInfoCache = 'enable-vm-info-cache',
+    HeartbeatDelayAllowance = 'heartbeat-delay-allowance',
+    HeartbeatInterval = 'heartbeat-interval',
+    HeartbeatLossCount = 'heartbeat-loss-count',
+    LicenseFileDirectory = 'license-file-directory',
+    PrimaryElectionTimeout = 'primary-election-timeout',
+    PrimaryScalingGroupName = 'primary-scaling-group-name',
+    PaygScalingGroupDesiredCapacity = 'scaling-group-desired-capacity',
```
[the remark this answers is not in the document]
Comment (JaydenLiang) [reply autoscale-setting.ts:31]: Terms in our code are generic terms that do not tie to any specific platform. We still use the abbr. PAYG and BYOL to refer the two pricing models.

File: autoscale-setting.ts [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+        booleanType: false
+    },
+    [AutoscaleSetting.PrimaryScalingGroupName]: {
+        keyName: AutoscaleSetting.PrimaryScalingGroupName,
+        description: 'The name of the primary auto scaling group.',
+        editable: false,
+        jsonEncoded: false,
+        booleanType: false
+    },
+    [AutoscaleSetting.PaygScalingGroupDesiredCapacity]: {
```
[the remark this answers is not in the document]
Comment (JaydenLiang) [reply autoscale-setting.ts:337]: Terms in our code are generic terms that do not tie to any specific platform. We still use the abbr. PAYG and BYOL to refer the two pricing models.

File: fortigate-autoscale/aws/aws-fortigate-autoscale-service.ts [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        const byolDesiredCapacity = Number(
+            settings.get(AwsFortiGateAutoscaleSetting.ByolScalingGroupDesiredCapacity).value
+        );
+        const byolMinSize = Number(
+            settings.get(AwsFortiGateAutoscaleSetting.ByolScalingGroupMinSize).value
+        );
+        const byolMaxSize = Number(
+            settings.get(AwsFortiGateAutoscaleSetting.ByolScalingGroupMaxSize).value
+        );
+        const paygGroupName = settings.get(AwsFortiGateAutoscaleSetting.PaygScalingGroupName).value;
```
[the remark this answers is not in the document]
Comment (JaydenLiang) [reply fortigate-autoscale/aws/aws-fortigate-autoscale-service.ts:150]: Terms in our code are generic terms that do not tie to any specific platform. We still use the abbr. PAYG and BYOL to refer the two pricing models.

File: fortigate-autoscale/aws/aws-fortigate-autoscale-service.ts [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+                    byolDesiredCapacity,
+                    byolMinSize,
+                    byolMaxSize
+                ).catch(error => {
+                    this.proxy.logForError('BYOL scaling group failed to start.', error);
+                    return false;
+                })
+            );
+        }
+        if (paygGroupName) {
```
[the remark this answers is not in the document]
Comment (JaydenLiang) [reply fortigate-autoscale/aws/aws-fortigate-autoscale-service.ts:188]: Terms in our code are generic terms that do not tie to any specific platform. We still use the abbr. PAYG and BYOL to refer to the two pricing models.

File: fortigate-autoscale/fortigate-autoscale-settings.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        description: 'The DNS name of the elastic load balancer for the FortiGate scaling groups.',
+        editable: false,
+        jsonEncoded: false,
+        booleanType: false
+    },
+    [FortiGateAutoscaleSetting.FortiGateInternalElbDns]: {
+        keyName: FortiGateAutoscaleSetting.FortiGateInternalElbDns,
+        description:
+            'The DNS name of the elastic load balancer for the scaling ' +
+            'groups of services protected by FortiGate.',
```
[the remark this answers is not in the document]
Comment (klc178) [reply fortigate-autoscale/fortigate-autoscale-settings.ts:84]:
```suggestion
            'The DNS name of the internal elastic load balancer ' +
            'used by the FortiGate Autoscale solution.',
```

File: fortigate-autoscale/fortigate-autoscale-settings.ts [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            'FortiGate to the protected services behind the load balancer.',
+        editable: true,
+        jsonEncoded: false,
+        booleanType: false
+    },
+    [FortiGateAutoscaleSetting.FortiGateTrafficProtocol]: {
+        keyName: FortiGateAutoscaleSetting.FortiGateTrafficProtocol,
+        description:
+            'The protocol for the traffic to be routed by the load balancer through ' +
+            'FortiGate to the protected services behind the load balancer.',
```
Comment (JaydenLiang) [inline fortigate-autoscale/fortigate-autoscale-settings.ts:132]:
```suggestion
            'FortiGates to the protected services behind the load balancer.',
```

File: fortigate-autoscale/fortigate-autoscale-settings.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        description: 'The interface the FortiGate uses for configuration synchronization.',
+        editable: true,
+        jsonEncoded: false,
+        booleanType: false
+    },
+    [FortiGateAutoscaleSetting.FortiGateTrafficPort]: {
+        keyName: FortiGateAutoscaleSetting.FortiGateTrafficPort,
+        description:
+            'The port number for load balancer to route traffic through ' +
+            'FortiGate to the protected services behind the load balancer.',
```
Comment (JaydenLiang) [inline fortigate-autoscale/fortigate-autoscale-settings.ts]:
```suggestion
            'FortiGates to the protected services behind the load balancer.',
```

Comment (JaydenLiang) [inline fortigate-autoscale/fortigate-autoscale-settings.ts]:
```suggestion
            'FortiGates to the protected services behind the load balancer.',
```

File: fortigate-autoscale/aws/aws-routing-egress-traffic-via-primary-vm-strategy.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        else {
+            if (this.env.primaryVm.networkInterfaces.length >= 1) {
+                [networkInterfaceId] = this.env.primaryVm.networkInterfaces
+                    .filter(eni => eni.index === 0)
+                    .map(eni => eni.id);
+            }
+            if (!networkInterfaceId) {
+                throw new Error(
+                    `No network interface found on the primary vm (id: ${this.env.primaryVm.id}).` +
+                        ' This is a fatal error and an impposible situation!'
```
[the remark this answers is not in the document]
Comment (klc178) [reply fortigate-autoscale/aws/aws-routing-egress-traffic-via-primary-vm-strategy.ts]: how is this outdated?

File: fortigate-autoscale/fortigate-autoscale-settings.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    [FortiGateAutoscaleSetting.FortiGateAdminPort]: {
+        keyName: FortiGateAutoscaleSetting.FortiGateAdminPort,
+        description: 'The port number for administrative login to FortiGate.',
+        editable: true,
+        jsonEncoded: false,
+        booleanType: false
+    },
+    [FortiGateAutoscaleSetting.FortiGateAutoscaleVirtualNetworkId]: {
+        keyName: FortiGateAutoscaleSetting.FortiGateAutoscaleVirtualNetworkId,
+        description: 'Virtual Network ID of the FortiGate Autoscale.',
```
[the remark this answers is not in the document]
Comment (klc178) [reply fortigate-autoscale/fortigate-autoscale-settings.ts:60]: This has not been applied yet.

File: fortigate-autoscale/fortigate-autoscale-settings.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        editable: false,
+        jsonEncoded: false,
+        booleanType: false
+    },
+    [FortiGateAutoscaleSetting.FortiGateAutoscaleVirtualNetworkCidr]: {
+        keyName: FortiGateAutoscaleSetting.FortiGateAutoscaleVirtualNetworkCidr,
+        description: 'Virtual Network CIDR of the FortiGate Autoscale.',
```
[the remark this answers is not in the document]
Comment (klc178) [reply fortigate-autoscale/fortigate-autoscale-settings.ts:67]: This has not been applied yet.

File: fortigate-autoscale/fortigate-autoscale-settings.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (klc178) [reply fortigate-autoscale/fortigate-autoscale-settings.ts:83]: This has not been applied yet.

**Comment by klc178** [comment]
Changes suggested in fortigate-autoscale/fortigate-autoscale-settings.ts for lines 60, 67, 83-84 ave not been applied yet.

**Comment by JaydenLiang** [comment]
> Changes suggested in fortigate-autoscale/fortigate-autoscale-settings.ts for lines 60, 67, 83-84 ave not been applied yet.

now they are done in https://github.com/fortinet/autoscale-core/pull/3/commits/e31489352415002a95dfc3e94f5e833c844299d8

File: fortigate-autoscale/fortigate-autoscale-settings.ts [hunk 13]
```diff
@@ -80,7 +80,8 @@ export const FortiGateAutoscaleSettingItemDictionary: SettingItemDictionary = {
         keyName: FortiGateAutoscaleSetting.FortiGateInternalElbDns,
         description:
             'The DNS name of the elastic load balancer for the scaling ' +
```
Comment (klc178) [inline fortigate-autoscale/fortigate-autoscale-settings.ts:82]: This line is no longer needed. It seems my previous suggestion only selected one of two lines.

Comment (JaydenLiang) [inline fortigate-autoscale/fortigate-autoscale-settings.ts:82]:
```suggestion
```

**Review by klc178 — APPROVED** [review]

**Review by jamie-pate — APPROVED** [review]

**Pull request merged**